Draft CIP - Delegateless Automation (#55)
global-synchronizer-foundation/cips: cip-0058/cip-0058.md

## CIP Delegateless Automation

<pre>
 CIP: N/A
 Title: Delegateless Automation
 Author: Julien Tinguely
 Comments-Summary: No comments yet
 Status: Draft
 Type: Governance
 Created: 2025-04-21
 Approved: N/A
 License: CC0-1.0
</pre>

## Abstract

This CIP proposes to replace the elected SV delegate mechanism to improve fault tolerance. 

> Instead of relying on a single delegate, we propose an approach where all SV applications attempt to exercise delegate choices. 

Reviewer: Would it be more clear to say "each SV Party attempts to..."?

Author: I think `SV application` highlights better the fact that delegate choices are attempted by the automation backend (no manual process involved).

Reviewer: ok, then let's say 

"we propose an approach where the backend automation for each SV application attempts"

## Motivation

A delegate performs critical network-related tasks, such as for example advancing mining rounds or executing accepted proposals.
In the current version, there is no automatic fault tolerance to mitigate issues that may occur on the node. 
SVs can vote on a new delegate, but this is a manual process, and the single point of failure is just moved to a different SV.
A concrete problem is rounds stop progressing when the delegate is down.
This has been observed multiple times on devnet and testnet and highlights the need for a better fault tolerance mechanism.

## Rationale

All SV applications attempting to complete tasks provides the right fault tolerance guarantees and introduces no correctness issues as commands are already all idempotent. 

Given that the current delegate-based choices do not require high throughput, e.g., round advancement happens only every 10 minutes,
random delays and backoffs are sufficient to minimize contention.

While more complex approaches to minimize contention would be possible, 
e.g., defining a clear order in which SVs should attempt to process a given task that is determined based on some task identifier, 

> the implementation complexity of that does not seem to be justified.

Reviewer: I think we need to make a stronger case that this "attempt with backoff" approach is not going to harm performance.

Author: I added a small paragraph just above to try to make a stronger case. What do you think?

Reviewer: We also need to make more clear, in the SV App section of the specification, that this is not happening in place of the normal 2/3 voting; please explain that this is just triggering an action that has already received a 2/3 vote in favor.

## Backwards compatibility

This new approach is only active when all SV operators have vetted the new governance dar file.

> All choices and backend code involved is backward compatible.

Reviewer: ```suggestion
All choices and backend code involved are backward compatible.
```

Daml will reject, with an authorization error, any transaction where the SV application tries to specify any party other than its own.

### SV Application

The SV application is modified such that all SVs attempt to submit the commands currently only being submitted by the delegate.

This introduces some potential for conflict: if multiple SV applications submit the same transaction at the same time, 
for example, all attempt to advance the mining rounds at the same time, only one of them will succeed and the others fail, that is,
they contend with each other. Importantly, this is not a correctness issue: The Daml code already handles this case; for example, it can successfully manage if the delegate crashes during a submission and tries to resubmit the same command. But the failed transactions introduce unnecessary load on the network.

To minimize the impact, each SV application adds a random delay before each attempt and retries with a random backoff.
A new metric named `splice_trigger_attempted_total` is added to monitor trigger failures caused by contention and help in choosing good delay and backoff parameters.

### SV UI

> In the current implementation, there is a dedicated page in the SV UI to elect a new delegate for the next round, 

Reviewer: ```suggestion
In the current implementation, there is a dedicated page in the SV UI used to elect a new delegate for the next round, 
```

## CIP Delegateless Automation

<pre>
 CIP: N/A
 Title: Delegateless Automation
 Author: Julien Tinguely
 Comments-Summary: No comments yet
 Status: Draft
 Type: Governance
 Created: 2025-04-21
 Approved: N/A
 License: CC0-1.0
</pre>

## Abstract

This CIP proposes to replace the elected SV delegate mechanism to improve fault tolerance. 

> Instead of relying on a single delegate, we propose an approach where the backend automation for each SV application attempts to exercise delegate choices. 

Reviewer: ```suggestion
Instead of relying on a single delegate, we propose an approach where the backend automation for each instance of the Super Validator application, (one instance operating on each of the different Super Validator nodes), attempts to exercise delegate choices. 
```

Each application will apply a random delay, tuned to minimize contention.

## Specification

### Daml

In the new implementation, all current choices on the `DsoRules` template having `dsoDelegate` as controller will introduce
a new optional argument named `sv` to be used instead. This allows all SVs to exercise the choice by specifying their own party.

> Daml will reject a transaction where they try to specify any party other than their own with an authorization error.

Reviewer: ```suggestion
Daml will reject, with an authorization error, any transaction where the SV application tries to specify any party other than its own.
```

### SV Application

The SV application is modified such that all SVs attempt to submit the commands currently only being submitted by the delegate.

This introduces some potential for conflict: if multiple SV applications submit the same transaction at the same time, 
for example, all attempt to advance the mining rounds at the same time, only one of them will succeed and the others fail, that is,

> they contend with each other. Importantly, this is not a correctness issue: The Daml code already has to handle this case, 
> e.g., if the delegate crashes during a submission and tries to resubmit the same command. But the failed transactions introduce unnecessary load on the network.

Reviewer: ```suggestion
they contend with each other. Importantly, this is not a correctness issue: The Daml code already handles this case; for example, it can successfully manage if the delegate crashes during a submission and tries to resubmit the same command. But the failed transactions introduce unnecessary load on the network.
```

To minimize the impact, each SV application adds a random delay before each attempt and retries with a random backoff.
A new metric named `splice_trigger_attempted_total` is added to monitor trigger failures caused by contention and help in choosing good delay and backoff parameters.

### SV UI

In the current implementation, there is a dedicated page in the SV UI to elect a new delegate for the next round, 
intended to unlock the system if a delegate comes down.
In the new implementation, delegate election and its dedicated page in the SV UI are fully removed and 
associated Daml choices cannot be executed anymore.

## Motivation

> A delegate performs critical network-related tasks, such as for example advancing mining rounds or triggering vote proposal executions when the required 2/3 voting thresholds is reached.

Author: @waynecollier-da is that precise enough?

Reviewer: yes

In the current version, there is no automatic fault tolerance to mitigate issues that may occur on the node. 
SVs can vote on a new delegate, but this is a manual process, and the single point of failure is just moved to a different SV.
A concrete problem is rounds stop progressing when the delegate is down.
This has been observed multiple times on devnet and testnet and highlights the need for a better fault tolerance mechanism.

## Rationale

All SV applications attempting to complete tasks provides the right fault tolerance guarantees and introduces no correctness issues as commands are already all idempotent. 

Given that the current delegate-based choices do not require high throughput, e.g., round advancement happens only every 10 minutes,
random delays and backoffs are sufficient to minimize contention.

Our approach does not degrade performance in practice, as it introduces only minimal overhead in SV app automation while significantly enhancing fault tolerance. 
Most tasks succeed on the first attempt, rendering the backoff mechanism largely inactive.

> Empirical benchmarks show that delegate transactions accounted for approximately 1% of total transactions on DevNet and 1.6% on MainNet, 

Author: @waynecollier-da let me know if you want a more precise description or if stating the percentages and adding the script in splice is enough.

Reviewer: good, thanks
